src/utils.js

```diff
@@ -157,17 +159,47 @@ export function parseGPTSingleSizeArray(singleSize) {
   }
 };
 
-exports.getTopWindowLocation = function () {
-  let location;
+export function getTopWindowLocation() {
+  if (exports.inIframe()) {
+    return getIframeParentLoc();
+  } else {
+    return window.location;
+  }
+}
+
+const getIframeParentLoc = function() {
+  let loc = window.location;
   try {
-    // force an exception in x-domain enviornments. #1509
-    window.top.location.toString();
-    location = window.top.location;
+    if (window.$sf) {
+      loc = window.document.referrer;
+    } else {
+      if (window.document.location && window.document.location.ancestorOrigins &&
+        window.document.location.ancestorOrigins.length >= 1) {
+        loc = window.document.location.ancestorOrigins[window.document.location.ancestorOrigins.length - 1];
+      } else if (window.document.location) {
+        // force an exception in x-domain environments. #1509
+        window.top.location.toString();
+        loc = getNonWebKitIframeParentLoc();
+      }
+    }
+    loc = parse(loc);
   } catch (e) {
-    location = window.location;
+    this.logMessage('getTopParentLoc failure', e);
   }
+  return loc;
+};
 
-  return location;
+const getNonWebKitIframeParentLoc = function() {
+  let referrerLoc = '';
+  let currentWindow;
+  do {
+    currentWindow = currentWindow ? currentWindow.parent : window;
+    if (currentWindow.document && currentWindow.document.referrer) {
+      referrerLoc = currentWindow.document.referrer;
```

Comment:
won't this return a string when an object is expected?

Reply:
Returns a string to `getIframeParentLoc` which then gets sent to `parse` which turns the string into the expected object.